### ISSUE TITLE
EC2 Auto discovery and WAN replication with public ip addresses. 

### ISSUE THREAD

**Issue by fuadm** [hazelcast/hazelcast#1379]
When we use EC2 Auto discovery, Hazelcast uses private ip addresses. But when we set WAN replication across regions it is not able to communicate over private IP addresses. An option is to force hazelcast to bind to public ip by setting 
<network>
<public-address>ec2-xx.us-west-1.compute.amazonaws.com:5701</public-address>
</network>

or 

<properties> 
<property name="hazelcast.local.localAddress">ec2-54-193-63-117.us-west-1.compute.amazonaws.com</property> 
<property name="hazelcast.local.publicAddress">ec2-54-193-63-117.us-west-1.compute.amazonaws.com</property> 
</properties>

but this requires
1. static setting of IP
2. it brakes the EC2 discovery as it uses the private ip's to discover. 

So both EC2 discovery and WAN replication needs to be fixed to make the process clean and seamless.

**Comment by jjongsma** [comment]
This is also a big issue for Docker containerization (which should maybe be a separately tracked issue). Unless you configure containers to net=host and explicitly tell Hazelcast to bind to the host's IP, Hazelcast listens on a Docker-local IP that is port-mapped to the host, and every connection request is rejected with the same "This node is not requested endpoint" error because the host's IP that other nodes connect to doesn't match what Hazelcast sees inside the container.

Host networking mode is a workaround, but requiring that and explicit IP ranges for Hazelcast to listen on is not scalable for a containerized environment. At minimum there should probably be an option to configure separate private and public node addresses, which is how Cassandra addresses this issue (listen vs. broadcast). I'd prefer disabling this check completely though - there are other ways to validate node identity if that is a concern (SSL certs, etc).

**Comment by sbuettner** [comment]
We are seeing the same issue as described by @jjongsma when trying to deploy Hazelcast inside AWS Elastic Beanstalk using Docker. 

```
2015-06-20 11:19:36.302  INFO 1 --- [thread-Acceptor] com.hazelcast.nio.tcp.SocketAcceptor     : [172.17.0.2]:5701 [dev] [3.5] Accepting socket connection from /10.0.1.209:41004
2015-06-20 11:19:36.303  INFO 1 --- [        cached5] c.h.nio.tcp.TcpIpConnectionManager       : [172.17.0.2]:5701 [dev] [3.5] Established socket connection between /172.17.0.2:5701
2015-06-20 11:19:36.303  WARN 1 --- [.IO.thread-in-1] c.h.nio.tcp.TcpIpConnectionManager       : [172.17.0.2]:5701 [dev] [3.5] Wrong bind request from Address[172.17.0.2]:5701! This node is not requested endpoint: Address[10.0.1.227]:5701
2015-06-20 11:19:36.304  INFO 1 --- [.IO.thread-in-1] com.hazelcast.nio.tcp.TcpIpConnection    : [172.17.0.2]:5701 [dev] [3.5] Connection [/10.0.1.209:41004] lost. Reason: Socket explicitly closed
```

**Comment by sbuettner** [comment]
Amazon provides a [meta-data service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) on each instance which seems to be also accessible from inside docker containers. This service can be used to get the public ip address of the ec2 instance. Since hazelcast already provides tight integration with aws it could use this service when a certain config flag has been set to get this address.

**Comment by mmedenjak** [comment]
There are docker issues encountering the same problem. Some of them:
https://github.com/hazelcast/hazelcast/issues/10801
https://github.com/hazelcast/hazelcast/issues/9219
We are working on adding a new SPI to 3.9 which will allow the instance to "discover" and set it's own public address without it explicitly being set by the user.

**Comment by mmedenjak** [comment]
@fuadm @imranbohoran @pveentjer @jjongsma @sbuettner there is a new SPI that has just been merged and which will be released in the 3.9 version. It will allow you to define the bind and public address when starting the hazelcast instance and if used correctly it can be used to avoid issues when forming a cluster.
For now you will have to use the SPI yourself and write an implementation which will fix your issue but we are planning on releasing implementations of our own which will be bundled into plugins such as the docker or AWS plugin for easier deployment.
Please check out the new SPI:
https://github.com/hazelcast/hazelcast/blob/3cede71cad1fe87312f0901ff77f903ed2d4383d/hazelcast/src/main/java/com/hazelcast/spi/MemberAddressProvider.java
Please create a new issue or reopen this one if this does not suit your use case.
